hal/halx86/apic/apic.c

```diff
@@ -800,6 +1370,36 @@ KfRaiseIrql(
     /* Return old IRQL */
     return OldIrql;
 }
+#endif
+
+KIRQL
+NTAPI
+KeGetCurrentIrql(VOID)
+{
+    /* Return the IRQL */
+    return KeGetPcr()->Irql;
+}
+
+VOID
+FASTCALL
+KfLowerIrql(_In_ KIRQL NewIrql)
+{
+    HalpLowerIrqlHardwareInterrupts(NewIrql);
+    HalpCheckForSoftwareInterrupt(NewIrql, 0);
+}
+
+KIRQL
+FASTCALL
+KfRaiseIrql(_In_ KIRQL NewIrql)
+{
+    PKPCR Pcr = KeGetPcr();
+    KIRQL OldIrql;
+
+    OldIrql = Pcr->Irql;
+    Pcr->Irql = NewIrql;
+
+    return OldIrql;
+}
```

Comment:
Our current APIC code writes a value to the TPR register here (see `ApicLowerIrql` and `ApicSetIrql` from above), and that is a correct behavior (aka "what Windows does").

What's going on here, and why have you changed that?

Reply:
Let's start by not using this code for amd64 now...
It is also a fact that the current x86 APIC HALs do not work.
Unlike my variant.
I think the statement "and that is a correct behavior (aka" what Windows does ")." is controversial. It is just more suitable for my code)